tests/shadow/test_shadow_class.py

```diff
@@ -0,0 +1,474 @@
+
+        state = shadow.global_snapshots()
+        assert state.shape == (10000, 2**wires, 2**wires)
+
+        state = np.mean(state, axis=0)
+        expected = np.exp(np.arange(2**wires) * 2j * np.pi / (2**wires)) / (2 ** (wires / 2))
+        expected = np.outer(expected, np.conj(expected))
+
+        assert qml.math.allclose(state, expected, atol=1e-1)
+
+    @pytest.mark.parametrize("wires", [1, 3])
+    @pytest.mark.parametrize("interface", ["autograd", "jax", "tf", "torch"])
+    @pytest.mark.parametrize("snapshots", [1, 100])
+    def test_subset_reconstruction_integer(self, wires, interface, snapshots):
+        """Test that the state reconstruction is correct for different numbers
+        of used snapshots"""
+        circuit = hadamard_circuit(wires, interface=interface)
+        bits, recipes = circuit()
+        shadow = ClassicalShadow(bits, recipes)
+
+        state = shadow.global_snapshots(snapshots=snapshots)
+        assert state.shape == (snapshots, 2**wires, 2**wires)
+
+    @pytest.mark.parametrize("wires", [1, 3])
+    @pytest.mark.parametrize("interface", ["autograd", "jax", "tf", "torch"])
+    def test_subset_reconstruction_iterable(self, wires, interface):
+        """Test that the state reconstruction is correct for different indices
+        of considered snapshots"""
+        circuit = hadamard_circuit(wires, interface=interface)
+        bits, recipes = circuit()
+        shadow = ClassicalShadow(bits, recipes)
+
+        # choose 1000 random indices
+        snapshots = np.random.choice(np.arange(10000, dtype=np.int64), size=1000, replace=False)
+        state = shadow.global_snapshots(snapshots=snapshots)
+        assert state.shape == (len(snapshots), 2**wires, 2**wires)
+
+        # check the results against obtaining the full global snapshots
+        expected = shadow.global_snapshots()
+        for i, t in enumerate(snapshots):
+            assert np.allclose(expected[t], state[i])
+
+    def test_large_state_warning(self, monkeypatch):
+        """Test that a warning is raised when a very large state is reconstructed"""
+        circuit = hadamard_circuit(17, shots=2)
+        bits, recipes = circuit()
+        shadow = ClassicalShadow(bits, recipes)
+
+        msg = "Querying density matrices for n_wires > 16 is not recommended, operation will take a long time"
+
+        with monkeypatch.context() as m:
+            # don't run the actual state computation since we only want the warning
+            m.setattr(np, "einsum", lambda *args, **kwargs: None)
+            m.setattr(np, "reshape", lambda *args, **kwargs: None)
+
+            with pytest.warns(UserWarning, match=msg):
+                shadow.global_snapshots()
+
+
+@pytest.mark.all_interfaces
+class TestExpvalEstimation:
+    """Test that the expval estimation is correct for a variety of observables"""
+
+    @pytest.mark.parametrize("interface", ["autograd", "jax", "tf", "torch"])
+    @pytest.mark.parametrize(
```

Comment:
Instead of parametrizing the observables I would suggest passing a list of observables. In the current way it is executing 100 000 shots for every observable which is not in the spirit of classical shadows.
I did this in the expval PR3, @eddddddy can you change it here?

Reply:
Changed